pandas/core/frame.py

```diff
@@ -3229,18 +3230,66 @@ def _setitem_array(self, key, value):
             indexer = key.nonzero()[0]
             self._check_setitem_copy()
             self.iloc[indexer] = value
+
         else:
             if isinstance(value, DataFrame):
                 check_key_length(self.columns, key, value)
                 for k1, k2 in zip(key, value.columns):
                     self[k1] = value[k2]
+
```

Comment:
really like to move this stuff out of here into another module, but i guess that's for another day